From e38ae1b00b3e08493259fa920131214c76be6a49 Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Tue, 22 Sep 2020 12:24:18 -0700
Subject: [PATCH 1642/1851] v4l: xilinx-gamma: Fix warnings for NULL_RETURNS

commit 273aec9f86626daa6378f46f5659901792919ebb from
https://github.com/Xilinx/linux-xlnx.git

This patch fix the warnings for the NULL_RETURNS type.

The actual warnings are:

For __xg_get_pad_format():
Event return_null: 	Explicitly returning null.

For xg_get_format():
Event returned_null: 	"__xg_get_pad_format" returns "NULL".
Event dereference: 	Dereferencing "__xg_get_pad_format(xg, cfg,
fmt->pad, fmt->which)", which is known to be "NULL".

For xg_set_format():
Event returned_null: 	"__xg_get_pad_format" returns "NULL".
Event var_assigned: 	Assigning: "__format" = "NULL" return value from
"__xg_get_pad_format"
Event dereference: 	Dereferencing "__format", which is known to be
"NULL".

Signed-off-by: Vishal Sagar <vishal.sagar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-gamma.c | 29 ++++++++++++++++----
 1 file changed, 24 insertions(+), 5 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-gamma.c b/drivers/media/platform/xilinx/xilinx-gamma.c
index d7996d0f34fa..f9aa4813df4e 100644
--- a/drivers/media/platform/xilinx/xilinx-gamma.c
+++ b/drivers/media/platform/xilinx/xilinx-gamma.c
@@ -115,15 +115,22 @@ __xg_get_pad_format(struct xgamma_dev *xg,
 		    struct v4l2_subdev_pad_config *cfg,
 		    unsigned int pad, u32 which)
 {
+	struct v4l2_mbus_framefmt *format;
+
 	switch (which) {
 	case V4L2_SUBDEV_FORMAT_TRY:
-		return v4l2_subdev_get_try_format(
-					&xg->xvip.subdev, cfg, pad);
+		format = v4l2_subdev_get_try_format(&xg->xvip.subdev, cfg,
+						    pad);
+		break;
 	case V4L2_SUBDEV_FORMAT_ACTIVE:
-		return &xg->formats[pad];
+		format = &xg->formats[pad];
+		break;
 	default:
-		return NULL;
+		format = NULL;
+		break;
 	}
+
+	return format;
 }
 
 static void xg_set_lut_entries(struct xgamma_dev *xg,
@@ -177,8 +184,14 @@ static int xg_get_format(struct v4l2_subdev *subdev,
 			 struct v4l2_subdev_format *fmt)
 {
 	struct xgamma_dev *xg = to_xg(subdev);
+	struct v4l2_mbus_framefmt *format;
+
+	format = __xg_get_pad_format(xg, cfg, fmt->pad, fmt->which);
+	if (!format)
+		return -EINVAL;
+
+	fmt->format = *format;
 
-	fmt->format = *__xg_get_pad_format(xg, cfg, fmt->pad, fmt->which);
 	return 0;
 }
 
@@ -190,6 +203,9 @@ static int xg_set_format(struct v4l2_subdev *subdev,
 	struct v4l2_mbus_framefmt *__format;
 
 	__format = __xg_get_pad_format(xg, cfg, fmt->pad, fmt->which);
+	if (!__format)
+		return -EINVAL;
+
 	*__format = fmt->format;
 
 	if (fmt->pad == XVIP_PAD_SINK) {
@@ -207,6 +223,9 @@ static int xg_set_format(struct v4l2_subdev *subdev,
 	fmt->format = *__format;
 	/* Propagate to Source Pad */
 	__format = __xg_get_pad_format(xg, cfg, XVIP_PAD_SOURCE, fmt->which);
+	if (!__format)
+		return -EINVAL;
+
 	*__format = fmt->format;
 	return 0;
 }
-- 
2.31.1

